### PR TITLE
[libc] Add type_traits tests

### PR DESCRIPTION
**Pull request by gchatelet**
This is not exhaustive for now but it provides a placeholder for `invoke_result` test mentioned in #65750.

### CODE REVIEW

**Comment by gchatelet** [comment]
The buildkite failure is unrelated to this patch.

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+class Class {};
+union Union {};
+struct Struct {};
+
+using UnqualObjectTypes = testing::TypeList<int, float, Class, Union, Struct>;
+
+TYPED_TEST(LlvmLibcTypeTraitsTest, add_lvalue_reference, UnqualObjectTypes) {
+
+  // non-ref cv, adds ref
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:24]: add a test for  `T*` ?

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+  EXPECT_TRUE((is_same_v<add_lvalue_reference_t<const void>, const void>));
+  EXPECT_TRUE(
+      (is_same_v<add_lvalue_reference_t<volatile void>, volatile void>));
+  EXPECT_TRUE((is_same_v<add_lvalue_reference_t<const volatile void>,
+                         const volatile void>));
+}
+
+TYPED_TEST(LlvmLibcTypeTraitsTest, add_pointer, UnqualObjectTypes) {
+  // object types -> pointer type
+  EXPECT_TRUE((is_same_v<add_pointer_t<T>, T *>));
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:51]: `T*` ?

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+  EXPECT_TRUE((is_same_v<add_rvalue_reference_t<void>, void>));
+  EXPECT_TRUE((is_same_v<add_rvalue_reference_t<const void>, const void>));
+  EXPECT_TRUE(
+      (is_same_v<add_rvalue_reference_t<volatile void>, volatile void>));
+  EXPECT_TRUE((is_same_v<add_rvalue_reference_t<const volatile void>,
+                         const volatile void>));
+}
+
+TEST(LlvmLibcTypeTraitsTest, bool_constant) {
+  EXPECT_TRUE((bool_constant<true>::value));
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:101]: `bool_constant_v` ?

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+  EXPECT_TRUE((is_same_v<decay_t<volatile int &>, int>));
+  EXPECT_TRUE((is_same_v<decay_t<const volatile int &>, int>));
+}
+
+// TODO enable_if
+
+TEST(LlvmLibcTypeTraitsTest, false_type) { EXPECT_FALSE((false_type::value)); }
+
+TEST(LlvmLibcTypeTraitsTest, integral_constant) {
+  EXPECT_EQ((integral_constant<int, 4>::value), 4);
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:130]: `_v` ?

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+using IntegralTypes =
+    testing::TypeList<bool, char, short, int, long, long long, unsigned char,
+                      unsigned short, unsigned int, unsigned long,
+                      unsigned long long>;
+
+TYPED_TEST(LlvmLibcTypeTraitsTest, is_arithmetic, IntegralTypes) {
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:138]: You probably want to test enums there.

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+  EXPECT_FALSE((is_arithmetic_v<Struct>));
+}
+
+TEST(LlvmLibcTypeTraitsTest, is_array) {
+  EXPECT_FALSE((is_array_v<int>));
+  EXPECT_FALSE((is_array_v<float>));
+  EXPECT_FALSE((is_array_v<Struct>));
+
+  EXPECT_TRUE((is_array_v<Class[]>));
+  EXPECT_TRUE((is_array_v<Union[4]>));
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:160]: test `T*` ?

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+TEST(LlvmLibcTypeTraitsTest, is_base_of) {
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:163]: you should add a test for `class E : public B {};`

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+  struct C : protected A {};
+  EXPECT_TRUE((is_base_of_v<A, C>));
+  EXPECT_FALSE((is_base_of_v<C, A>));
+
+  struct D : private A {};
+  EXPECT_TRUE((is_base_of_v<A, D>));
+  EXPECT_FALSE((is_base_of_v<D, A>));
+}
+
+TEST(LlvmLibcTypeTraitsTest, is_class) {
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:180]: What about `enum class`, `T*` and `T&`?

File: libc/test/src/__support/CPP/type_traits_test.cpp [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+  EXPECT_TRUE((is_class_v<Struct>));
+  EXPECT_TRUE((is_class_v<Class>));
+  EXPECT_FALSE((is_class_v<Union>));
+  EXPECT_FALSE((is_class_v<int>));
+}
+
+TYPED_TEST(LlvmLibcTypeTraitsTest, is_const, UnqualObjectTypes) {
```
Comment (legrosbuffle) [inline libc/test/src/__support/CPP/type_traits_test.cpp:187]: add a test for
```
using Alias = const T;
EXPECT_TRUE((is_const_v<Alias>));
```

File: libc/test/src/__support/CPP/type_traits_test.cpp (hunk 3 above)
Comment (legrosbuffle) [reply libc/test/src/__support/CPP/type_traits_test.cpp:116]: Interesting...

**Review by legrosbuffle — APPROVED** [review]

**Pull request merged**